src/platform/intel/ace/lib/watchdog.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+// Author: Adrian Warecki <adrian.warecki@intel.com>
+
+#include <sof/lib/uuid.h>
+#include <sof/drivers/idc.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <sof/ipc/msg.h>
+#include <ipc4/notification.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/watchdog.h>
+
+#include <adsp_watchdog.h>
+#include <intel_adsp_ipc.h>
+#include <intel_adsp_ipc_devtree.h>
+
+LOG_MODULE_REGISTER(wdt, CONFIG_SOF_LOG_LEVEL);
+
+/* 13c8bc59-c4fa-4ad1-b93a-ce97cd30acc7 */
+DECLARE_SOF_UUID("wdt", wdt_uuid, 0x13c8bc59, 0xc4fa, 0x4ad1,
+		 0xb9, 0x3a, 0xce, 0x97, 0xcd, 0x30, 0xac, 0xc7);
+
+DECLARE_TR_CTX(wdt_tr, SOF_UUID(wdt_uuid), LOG_LEVEL_INFO);
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adsp_watchdog), okay)
+const struct device *const watchdog = DEVICE_DT_GET(DT_NODELABEL(adsp_watchdog));
+struct ipc_msg slave_timeout_ipc;
+
+void wdt_slave_core_action_on_timeout()
+{
+	struct idc_msg msg;
+
+	msg.header = IDC_MSG_SLAVE_CORE_CRASHED | IDC_SCC_CORE(cpu_get_id()) |
+		     IDC_SCC_REASON(IDC_SCC_REASON_WATCHDOG);
+	msg.extension = 0;
+	msg.core = 0;
+	msg.size = 0;
+	msg.payload = NULL;
+	idc_send_msg(&msg, IDC_NON_BLOCKING);
+}
+
+void wdt_master_core_action_on_timeout()
+{
+	struct ipc4_watchdog_timeout_notification notif;
+
+	/* Send Watchdog Timeout IPC notification */
+	ipc4_notification_watchdog_init(&notif, cpu_get_id(), true);
+	intel_adsp_ipc_send_message_emergency(INTEL_ADSP_IPC_HOST_DEV,
+					      notif.primary.dat, notif.extension.dat);
+}
+
+static void wdt_timeout(const struct device *dev, int core)
+{
+	if (!core)
+		wdt_master_core_action_on_timeout();
+	else
+		wdt_slave_core_action_on_timeout();
+}
+
+void watchdog_init(void)
```

Comment:
would it be better to have all functions in this file use the same naming - either `watchdog_*` or `wdt_*`?